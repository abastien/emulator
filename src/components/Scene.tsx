import React, {useEffect, useLayoutEffect, useRef, useState} from "react";

const Scene = (): JSX.Element => {
  const [width, setWidth] = useState(window.innerWidth);
  const [height, setHeight] = useState(window.innerHeight);
  const canvasRef = useRef(null);

  useLayoutEffect(() => {
    const canvasObj = canvasRef.current;
    const canvasCtx = canvasObj.getContext('2d');
    canvasCtx.fillStyle = 'black';
    canvasCtx.fillRect(0, 0, width, height);
  }, [width, height]);

  useEffect(() => {
    const resize = (): void => {
      setWidth(window.innerWidth);
      setHeight(window.innerHeight);
    }
    window.addEventListener('resize', resize);
    return () => {
      console.log('Scene cleanup: Detach resize event listener');
      window.removeEventListener('resize', resize);
    }
  }, [])

  return <canvas ref={canvasRef} width={width} height={height} />;
}

export default Scene;
