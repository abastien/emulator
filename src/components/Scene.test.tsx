import React from 'react';
import {
  cleanup, render, fireEvent, act,
} from '@testing-library/react';
import Scene from './Scene';

interface CustomWindow extends Window {
  innerWidth: number
  innerHeight: number
}

describe('Scene', () => {
  const defaultWidth = 1024;
  const defaultHeight = 768;
  const mockWindow: CustomWindow = window;

  beforeEach(() => {
    mockWindow.innerWidth = defaultWidth;
    mockWindow.innerHeight = defaultHeight;
  });

  test('should render a canvas', () => {
    /** given */
    /** when */
    const { asFragment } = render(<Scene />);
    /** then */
    expect(asFragment()).toMatchSnapshot();
  });

  test('should adapt the canvas to window size', () => {
    /** given */
    /** when */
    const { baseElement } = render(<Scene />);
    /** then */
    const elements = baseElement.getElementsByTagName('canvas');
    expect(elements.length).toEqual(1);
    expect(elements[0].width).toEqual(defaultWidth);
    expect(elements[0].height).toEqual(defaultHeight);
  });

  test('should resize the canvas on window resize', () => {
    /** given */
    const customWidth = 1600;
    const customHeight = 1200;
    /** when */
    const { baseElement } = render(<Scene />);
    act(() => {
      mockWindow.innerWidth = customWidth;
      mockWindow.innerHeight = customHeight;
      fireEvent(mockWindow, new Event('resize'));
    });
    /** then */
    const elements = baseElement.getElementsByTagName('canvas');
    expect(elements.length).toEqual(1);
    expect(elements[0].width).toEqual(customWidth);
    expect(elements[0].height).toEqual(customHeight);
  });

  afterAll(cleanup);
});
