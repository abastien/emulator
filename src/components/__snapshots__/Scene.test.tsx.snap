// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene should render a canvas 1`] = `
<DocumentFragment>
  <canvas
    height="768"
    width="1024"
  />
</DocumentFragment>
`;
